prog/dftb+/lib_type/typegeometryhsd.F90

```diff
@@ -429,6 +417,231 @@ subroutine readTGeometryXyz_help(node, geo, text)
   end subroutine readTGeometryXyz_help
 
 
+  !> Reads the geometry from a node in a HSD tree in GEN format
+  subroutine readTGeometryVasp(node, geo)
+
+    !> Node containing the geometry in Gen format
+    type(fnode), pointer :: node
+
+    !> Contains the geometry on exit
+    type(TGeometry), intent(out) :: geo
+
+    type(string) :: text
+
+    call getFirstTextChild(node, text)
+    call readTGeometryVasp_help(node, geo, char(text))
+
+  end subroutine readTGeometryVasp
+
+
+  !> Helping routine for reading geometry from a HSD tree in VASP format
+  subroutine readTGeometryVasp_help(node, geo, text)
+
+    !> Node to parse (only needed to produce proper error messages)
+    type(fnode), pointer :: node
+
+    !> Contains the geometry on exit
+    type(TGeometry), intent(out) :: geo
+
+    !> Text content of the node
+    character(len=*), intent(in) :: text
+
+    type(string) :: txt
+    character(mc), allocatable :: vaspNames(:)
+    integer :: iStart, iOldStart, iErr, iEnd
+    integer :: ii, jj, iSp, iTmp
+    real(dp) :: coords(3), latVec(3), rTmp, rScale
+    integer, allocatable :: vaspSp(:)
+    integer, allocatable :: countSp(:)
+    type(listString) :: speciesNames
+    logical :: hasComment
+
+
+    ! Read `first' line of the POSCAR/CONTCAR file
+    ! This is actually a comment line, but contains by user convention the atomic symbols
+    ! In case it does not contain the atomic symbols, is empty or an HSD comment,
+    ! we have to ignore it, therefore we attempt to read it as `second' line first
+    hasComment = .false.
+    iStart = 1
+    iEnd = nextLine(text, iStart)
+    print'(a)', text(iStart:iEnd-1)
+    call getNextToken(text(:iEnd), rScale, iStart, iErr)
+    ! seems like we found the `second' line already, to be sure we check for more token
+    if (iErr == TOKEN_OK) then
+      hasComment = iStart <= iEnd
+    else
+      hasComment = .true.
+    end if
+
+    if (hasComment) then
+      iStart = 1
+      call init(speciesNames)
+      iErr = TOKEN_OK
+      do while(iErr == TOKEN_OK)
+        call getNextToken(text(:iEnd), txt, iStart, iErr)
+        if (iErr == TOKEN_OK) then
+          call append(speciesNames, char(txt))
+        end if
+      end do
+      iStart = iEnd + 1
+
+      ! try to read the real `second' line now
+      iEnd = nextLine(text, iStart)
+      call getNextToken(text(:iEnd), rScale, iStart, iErr)
+      call checkError(node, iErr, "Bad scaling factor.")
```

Comment:
Its more user-friendly if the error is explicit as to what the user should fix/is wrong in the  input:
"Bad scaling factor on line 2 of the VASP geometry"

Reply:
Changed.